tls/s2n_security_rules.h

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include "stuffer/s2n_stuffer.h"
+#include "utils/s2n_result.h"
+
+typedef enum {
+    S2N_PERFECT_FORWARD_SECRECY = 0,
+    S2N_SECURITY_RULES_COUNT,
+} s2n_security_rule_id;
+
+#define TO_FLAG(rule) rule##_FLAG = (1 << (rule))
+typedef enum {
+    TO_FLAG(S2N_PERFECT_FORWARD_SECRECY),
+} s2n_security_flag;
+
+struct s2n_security_rule_result {
+    bool found_error;
+    bool output_truncated;
+    struct s2n_stuffer output;
+};
+S2N_CLEANUP_RESULT s2n_security_rule_result_free(struct s2n_security_rule_result *result);
+
+struct s2n_security_policy;
+struct s2n_cipher_suite;
+struct s2n_signature_scheme;
+struct s2n_ecc_named_curve;
+
+struct s2n_security_rule {
+    const char *name;
+    S2N_RESULT (*validate_cipher_suite)(const struct s2n_cipher_suite *cipher_suite, bool *valid);
+    S2N_RESULT (*validate_sig_scheme)(const struct s2n_signature_scheme *sig_scheme, bool *valid);
+    S2N_RESULT (*validate_cert_sig_scheme)(const struct s2n_signature_scheme *sig_scheme, bool *valid);
+    S2N_RESULT (*validate_curve)(const struct s2n_ecc_named_curve *curve, bool *valid);
```

Comment:
Ultimately I want to use these for more than just security policies. For example, I want to check the actual security parameters chosen by the connection for FIPS compliance.

If we run into a rule that requires checking multiple policy fields, we can reconsider. But I suspect that won't be common, so I'd like to try to keep these simple.

Reply:
Like, if you have a `validate_cipher_suite(cipher, is_valid)` method, it's going to be doing pretty simple checks and we're probably fine with the common security rule code handling logging the error to output just based on "is_valid". If you're going to do more complex checks though, we'll want to pass the result object too so that you can do your own logging, and now your implementation of your rule is getting more complicated :)

Reply:
Do we think these callbacks are going to scale with more complex usecases? For example, let's say a ciphersuite should only be used with a specific chosen version of TLS. Or a ciphersuite and curve should be chosen together.

Reply:
No, but I don't think that's an issue.

Both your hypotheticals are only relevant to checking a connection / handshake, not a security policy. A security policy can't assume which version of TLS will be chosen (it can check the policy's min version, but that's not really sufficient) or guarantee that two options (a cipher suite and a curve) will be chosen together. Like I call out in the description, we'd probably want a `validate_handshake(struct s2n_connection *conn, struct s2n_security_rule_result *result)` method to cover cases like that when specifically checking the actual negotiated parameters instead of a security policy. We also don't have any cases for checks like that yet-- both FIPS and RFC5191 don't need that. And even if they did, they'd likely still rely heavily on much simpler checks that can be more clearly implemented with these simple callbacks.